langchain/src/tools/discord.ts

```diff
@@ -0,0 +1,377 @@
+    await this.client.destroy();
+    const results =
+      messages.map((message: Message) => ({
+        author: message.author.tag,
+        content: message.content,
+        timestamp: message.createdAt,
+      })) ?? [];
+
+    return JSON.stringify(results);
+  }
+}
+
+/**
+ * A tool for retrieving all servers a bot is a member of. It extends the
+ * base `Tool` class and implements the `_call` method to perform the retrieve
+ * operation. Requires a bot token which can be set in the environment
+ * variables.
+ */
+export class DiscordGetGuildsTool extends Tool {
+  static lc_name() {
+    return "DiscordGetGuildsTool";
+  }
+
+  name = "discord-get-guilds";
+
+  description = `a discord tool. useful for getting a list of all servers/guilds the bot is a member of. no input required.`;
+
+  protected botToken: string;
+
+  client = new Client({
+    intents: [GatewayIntentBits.Guilds],
+  });
+
+  constructor(fields: DiscordToolParams = {}) {
+    super();
+
+    const { botToken = getEnvironmentVariable("DISCORD_BOT_TOKEN") } =
+      fields || {};
+
+    if (!botToken) {
+      throw new Error(
+        "Discord API key not set. You can set it as DISCORD_BOT_TOKEN"
+      );
+    }
+    this.botToken = botToken;
+  }
+
+  /** @ignore */
+  async _call(_input: string): Promise<string> {
+    await this.client.login(this.botToken);
+
+    const guilds = await this.client.guilds.fetch();
+    await this.client.destroy();
+
+    const results =
+      guilds.map((guild) => ({
+        id: guild.id,
+        name: guild.name,
+        createdAt: guild.createdAt,
+      })) ?? [];
+
+    return JSON.stringify(results);
+  }
+}
+
+/**
+ * A tool for retrieving text channels within a server/guild a bot is a member
+ * of. It extends the base `Tool` class and implements the `_call` method to
+ * perform the retrieve operation. Requires a bot token which can be set in
+ * the environment variables. The `_call` method takes a server/guild ID
+ * to get its text channels.
+ */
+export class DiscordGetTextChannelsTool extends Tool {
+  static lc_name() {
+    return "DiscordGetTextChannelsTool";
+  }
+
+  name = "discord-get-text-channels";
+
+  description = `a discord tool. useful for getting a list of all text channels in a server/guild. input should be a discord server/guild ID`;
+
+  protected botToken: string;
+
+  client = new Client({
+    intents: [GatewayIntentBits.Guilds],
+  });
+
+  constructor(fields: DiscordToolParams = {}) {
+    super();
+
+    const { botToken = getEnvironmentVariable("DISCORD_BOT_TOKEN") } =
+      fields || {};
+
+    if (!botToken) {
+      throw new Error(
+        "Discord API key not set. You can set it as DISCORD_BOT_TOKEN"
+      );
+    }
+    this.botToken = botToken;
+  }
+
+  /** @ignore */
+  async _call(input: string): Promise<string> {
+    await this.client.login(this.botToken);
+
+    const guild = await this.client.guilds.fetch(input);
+    const channels = await guild.channels.fetch();
```

Comment:
I vaguely remember there being two methods for getting content from discord, one which used a cache and one which didn't. Just want to make sure you're aware of this and are using the _non cached_ method.